### PR TITLE
Cross-version API Compatibility Testing- Server APIs

### PR DESCRIPTION
**Pull request by mpanchajanya**
### What this PR does / why we need it

- Added Server APIs tests on supported Runtime library versions v0.11.6, v0.25.4, v0.28.0, latest.
- Added helper methods to create API Commands, Validators to validate input and outputOptions
- Extended the testplugins implementation to trigger Server APIs.
- Set/Get/ DeleteServer, Set/Get/RemoveCurrentServer API methods are tested for supported Runtime Library versions 1.0.0/latest, 0.28.0, 0.25.4, 0.11.6.
- Added group of tests using single server involving Set/Get/ DeleteServer, Set/Get/RemoveCurrentServer API methods.
- Added group of tests using two servers involving Set/Get/ DeleteServer, Set/Get/RemoveCurrentServer API methods.

Implemented  Helper methods to create commands and input & output options in the framework.

### Which issue(s) this PR fixes

Fixes #

### Describe testing done for PR

Added Unit testing for the helper functions like Command creations, Trigger Server APIs.

Commands
Build and Run compatibility tests
`make compatibility-tests`

Build all the runtime specific version test plugins. Binaries of all the test plugins needs to be build to run any compatibility test.
`make build-compatibility-test-plugins`

Run all compatibility tests
`make run-compatibility-tests`

What will the tests cover ?
Interactions of different implementations of APIs across various versions of library works as expected.

What will be the test results ?
GitHub CI runner pipeline include Logs with details on test cases from test-case that are executed and succeeded/failed.

For more details on Cross-version API compatibility testing refer to https://github.com/vmware-tanzu/tanzu-plugin-runtime/blob/main/test/compatibility/docs/cross-version-api-compatibility.md

### Release note

```release-note
Added Cross-version Server APIs Compatibility tests
```

### CODE REVIEW

File: test/compatibility/framework/compatibilitytests/common/test_helpers.go [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+// Package common contains all the cross version api compatibility tests
+package common
+
+const (
+	DefaultEndpoint = "default-compatibility-test-endpoint"
+)
+
+const (
+	CtxCompatibilityOne string = "compatibility-one"
```
Comment (mpanchajanya) [inline test/compatibility/framework/compatibilitytests/common/test_helpers.go:12]: Renamed to be more generic in follow up PR

File: test/compatibility/framework/compatibilitytests/server/server_test.go [hunk 2]
```diff
@@ -0,0 +1,751 @@
...
+		// Setup mock temporary config files for testing
+		_, cleanup := core.SetupTempCfgFiles()
+		ginkgo.DeferCleanup(func() {
+			cleanup()
+		})
+	})
+
+	ginkgo.Context("using single server", func() {
+
+		ginkgo.It("SetServer, SetCurrentServer latest then GetServer, GetCurrentServer v0.25.4, v0.28.0, latest then DeleteServer, RemoveCurrentServer v0.28.0 then GetServer, GetCurrentServer v0.25.4, v0.28.0, latest", func() {
```
Comment (mpanchajanya) [inline test/compatibility/framework/compatibilitytests/server/server_test.go:34]: Refining the test case with resusable options and commands

File: test/compatibility/framework/compatibilitytests/server/server_test.go [hunk 3]
```diff
@@ -0,0 +1,751 @@
...
+			// Input and Output Parameters for GetCurrentServer
+			getCurrentServerInputOptionsForRuntimeLatest := server.DefaultGetCurrentServerInputOptions(core.VersionLatest)
+			getCurrentServerInputOptionsForRuntime0280 := server.DefaultGetCurrentServerInputOptions(core.Version0280)
+			getCurrentServerInputOptionsForRuntime0254 := server.DefaultGetCurrentServerInputOptions(core.Version0254)
+
+			getCurrentServerOutputOptionsForRuntimeLatest := server.DefaultGetCurrentServerOutputOptions(core.VersionLatest, common.CtxCompatibilityOne)
+			getCurrentServerOutputOptionsForRuntime0280 := server.DefaultGetCurrentServerOutputOptions(core.Version0280, common.CtxCompatibilityOne)
+			getCurrentServerOutputOptionsForRuntime0254 := server.DefaultGetCurrentServerOutputOptions(core.Version0254, common.CtxCompatibilityOne)
+
+			//getCurrentServerOutputOptionsForRuntimeLatestWithError := server.DefaultGetCurrentServerOutputOptionsWithError(core.VersionLatest)
```
Comment (chandrareddyp) [inline test/compatibility/framework/compatibilitytests/server/server_test.go:165]: if there is reason to keep this commented code, then add TODO or comment

File: test/compatibility/framework/compatibilitytests/server/server_test.go
Comment (chandrareddyp) [inline test/compatibility/framework/compatibilitytests/server/server_test.go:1]: If there is a reason to keep this commented code at lines 165,177, 206 etc, then add TODO or comment

File: test/compatibility/framework/compatibilitytests/server/server_test.go [hunk 4]
```diff
@@ -0,0 +1,751 @@
...
+			// Add DeleteServer v0.28.0 Command
+			testCase.Add(deleteCtxCmd)
+
+			// Add GetServer latest, v0.28.0, v0.25.4 Commands
+			testCase.Add(getServerCmdForRuntimeLatestWithError).Add(getServerCmdForRuntime0280WithError).Add(getServerCmdForRuntime0254WithError)
+
+			// Add GetCurrentServer latest, v0.28.0, v0.25.4 Commands
+			testCase.Add(getCurrentServerCmdForRuntimeLatestWithError).Add(getCurrentServerCmdForRuntime0280WithError).Add(getCurrentServerCmdForRuntime0254WithError)
+
+			ginkgo.By("Execute the testcase")
```
Comment (chandrareddyp) [inline test/compatibility/framework/compatibilitytests/server/server_test.go:143]:
```suggestion
			ginkgo.By("Execute the test case")
```

File: test/compatibility/framework/compatibilitytests/server/server_test.go [hunk 5]
```diff
@@ -0,0 +1,751 @@
...
+
+			// Run all the commands
+			executer.Execute(testCase)
+		})
+		ginkgo.It("SetServer, SetCurrentServer v0.25.4 then GetServer, GetCurrentServer v0.25.4, v0.28.0, latest then DeleteServer, RemoveCurrentServer v0.28.0 then GetServer, GetCurrentServer v0.25.4, v0.28.0, latest", func() {
```
Comment (chandrareddyp) [inline test/compatibility/framework/compatibilitytests/server/server_test.go:148]: can you check the sequence, seems its not in sync with the commands at line 238

Comment (mpanchajanya) [reply test/compatibility/framework/compatibilitytests/server/server_test.go:148]: Refactored the test sequence

File: test/compatibility/framework/compatibilitytests/server/server_test.go (hunk 4 above)
Comment (mpanchajanya) [reply test/compatibility/framework/compatibilitytests/server/server_test.go:143]: Updated

File: test/compatibility/framework/compatibilitytests/server/server_test.go
Comment (mpanchajanya) [reply test/compatibility/framework/compatibilitytests/server/server_test.go:1]: Cleaned the test cases

**Comment by mpanchajanya** [comment]
> lgtm
> 
> question: (I might have asked and forgotten) any specific reason why the compat tests are not wired up to the test target? They don't seem that expensive to run.

Actually it is wired to run on pull_requests to main and release-*

**Pull request merged**